
# RNA-forge (version 1.1)
This repository contains the official implementation of our method RNA-forge: Ultra-fast, Alignment-free *Ab initio* RNA Tertiary Structure Prediction

## The architecture of RNA-forge

From an input RNA sequence, embeddings are generated by a fine-tuned RNA language model and passed through a symmetric convolutional head to predict an inter-nucleotide phosphorus distance matrix. Classical MDS then constructs the initial P-backbone 3D coordinates. Using these coordinates together with predicted torsion angles and sampled bond geometries, the BuildIter module applies torsional rotations and rigid alignments of sugars and bases to assemble the full-atom RNA structure.

![RNA-forge Pipeline](images/rna_forge_pipeline.png)


## Installation

### For Linux
(MacOS is not currently supported)

### Environment Setup

Create and activate a Python virtual environment:

```bash
python3 -m venv forge-venv
source forge-venv/bin/activate
```

### Dependencies

Install required packages using:

```bash
pip install -r requirements.txt
```

### RNA-TorsionBERT Setup

Clone and setup RNA-TorsionBERT in the project directory:

```bash
git clone https://github.com/EvryRNA/RNA-TorsionBERT.git
cd RNA-TorsionBERT
git checkout e901a6b9a84051d643248c405be2ebd8af95e1a9
pip install -r requirements.txt
cd ..
```

### Pre-trained Models

Download the pre-trained models from [Google Drive](https://drive.google.com/drive/folders/1r3FoHxolCXdw5lyepiZbx9HkWklNp3WZ?usp=sharing) and place them in the `trained_models/` folder.

## Usage

### Basic Usage

Predict RNA 3D structure from a FASTA file:

```bash
python predict_struct.py --input_fasta=test/test.fasta --output_pdb_path=test/out.pdb
```

### Command Line Options

| Option                        | Choices                | Default      | Description                                      |
| ----------------------------- | ---------------------- | ------------ | ------------------------------------------------ |
| `--input_fasta`               | -                      | **Required** | Path to input FASTA file containing RNA sequence |
| `--distance_matrix_predictor` | `rinalmo`, `birnabert` | `birnabert`  | Model to use for distance matrix prediction      |
| `--torsion_angle_predictor`   | `rtb`, `drt`           | `rtb`        | Model to use for torsion angle prediction.       |
| `--output_pdb_path`           | -                      | **Required** | Path to save the output PDB file                 |
| `--seed`                      | integer                | `42`         | Random seed used for sampling                    |


## Project Structure

- `training_notebooks/` - Contains training notebook for the distance matrix predictor models
- `training_sequences/` - Training and validation sequences used for model training
- `data_helper/` - Auxiliary data files, including bond geometry parameters and template PDB structures utilized by the BuildIter module.
- `trained_models/` - Pre-trained model files are kept here
- `test/` - Sample test fasta file

## Additional Information
The [Google Drive](https://drive.google.com/drive/folders/1r3FoHxolCXdw5lyepiZbx9HkWklNp3WZ?usp=sharing) link also contains the processed training data used for model training.

## Note
The current version does not include the pretrained model for DeepRNA-Twist; however, RNA-TorsionBERT remains fully functional within the existing RNA-forge pipeline.
